include/deal.II/base/config.h.in

```diff
@@ -173,6 +173,20 @@
 #endif
 #cmakedefine DEAL_II_FALLTHROUGH @DEAL_II_FALLTHROUGH@
 #cmakedefine DEAL_II_CONSTEXPR @DEAL_II_CONSTEXPR@
+// clang++ assumes that all constexpr functions are __host__ __device__ when
+// compiling CUDA code, i.e, when Kokkos was configured with CUDA support.
+// This is problematic when calling non-constexpr functions in constexpr
+// functions. Also, we intend constexpr functions to only be device-callable if
+// not marked with DEAL_II_HOST_DEVICE. Hence, just annotate constexpr
+// functions additionally with "__host__".
+#if defined(DEAL_II_CONSTEXPR)
+#  if defined(__clang__) && defined(__CUDA__)
+#    define DEAL_II_CONSTEXPR_FUNCTION DEAL_II_CONSTEXPR __host__
+#  else
+#    define DEAL_II_CONSTEXPR_FUNCTION DEAL_II_CONSTEXPR
+#  endif
+#endif
```

Comment:
Wouldn't it be more obvious if we marked these functions up where they are defined, rather than through the backdoor of this macro?

Reply:
That means that I need to have this annotation for every single `constexpr` function in the library. To me, that seems more of a nuisance than defining a separate macro. Or am I missing something?

Reply:
Yes, but you've already done all of the work identifying which of the functions need to be annotated. Just replace the `#define` by its components.

The way I see it is that the new `#define` you're introducing here is mixing two different aspects. Whether a function is `constexpr` and/or `__host__` is two different things, and they can appear in any combination. At least that is my understanding (I'm willing to be corrected). If so, it is confusing if there is one `#define` that picks one of the possible combinations -- let's let each thing be responsible for one concern.

Reply:
I basically need the annotation anywhere a `constexpr` function calls a `host`-only function. On the other hand, I have only changed functions in a file the compiler was complaining about and I didn't try to find all `constexpr` functions. I see that this looks a little inconsistent now and I would be happy to apply the macro to all `constexpr` functions. Would you find that acceptable?

Reply:
@bangerth I added a commit that replaces the `#define`.

Reply:
I do not think this is pretty, but I think the current version is acceptable (and makes it clear how to fix up in case the problem disappears in the future).